File: js/core/src/tracing/instrumentation.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  ROOT_CONTEXT,
  trace,
  TraceFlags,
  type Span as ApiSpan,
  type Link,
} from '@opentelemetry/api';
import { performance } from 'node:perf_hooks';
import { getAsyncContext } from '../async-context.js';
import type { HasRegistry, Registry } from '../registry.js';
import { ensureBasicTelemetryInstrumentation } from '../tracing.js';
import { DirectTelemetryInstrumentation } from './direct-instrumentation.js';
import { telemetryServerUrl } from './exporter.js';
import {
  ATTR_PREFIX,
  firstNonEmpty,
  metadataToAttributes,
  SPAN_TYPE_ATTR,
  spanMetadataAlsKey,
  TRACER_NAME,
  TRACER_VERSION,
  type GenkitSpanContext,
  type Instrumentation,
  type InstrumentationSpanInfo,
} from './instrumentation-api.js';
import { OTelInstrumentation } from './otel-instrumentation.js';
import type { PathMetadata, SpanMetadata, TraceMetadata } from './types.js';

export { DirectTelemetryInstrumentation } from './direct-instrumentation.js';
export {
  ATTR_PREFIX,
  disableOTelRootSpanDetection,
  SPAN_TYPE_ATTR,
  spanMetadataAlsKey,
  type DisposableInstrumentation,
  type GenkitLogRecord,
  type GenkitSpanContext,
  type Instrumentation,
  type InstrumentationSpanInfo,
  type LogRecordingInstrumentation,
} from './instrumentation-api.js';
export { OTelInstrumentation } from './otel-instrumentation.js';

type SpanContext = {
  metadata: SpanMetadata;
  labels?: Record<string, string>;
  spanId?: string;
} & TraceMetadata;

interface RunInNewSpanOpts {
  metadata: SpanMetadata;
  labels?: Record<string, string>;
  links?: Link[];
}

type RunInNewSpanFn<T> = (
  metadata: SpanMetadata,
  otSpan: ApiSpan
) => Promise<T>;

// ---------------------------------------------------------------------------
// Provider registry (global-keyed so it survives module duplication).
// ---------------------------------------------------------------------------

const configuredInstrumentationKey = '__GENKIT_CONFIGURED_INSTRUMENTATION';
const directInstrumentationKey = '__GENKIT_DIRECT_INSTRUMENTATION';
const defaultOTelInstrumentationKey = '__GENKIT_DEFAULT_OTEL_INSTRUMENTATION';

/**
 * Replaces the implicit default (`OTelInstrumentation`) with the provided
 * instrumentation. `DirectTelemetryInstrumentation` is still prepended in dev
 * (when a telemetry server is configured), so a typical dev chain becomes
 * `[Direct, x]`.
 *
 * @hidden
 */
export function configureInstrumentation(i: Instrumentation) {
  global[configuredInstrumentationKey] = i;
}

/**
 * Clears configured/auto-injected instrumentation. Tests only.
 *
 * @hidden
 */
export function resetInstrumentation() {
  delete global[configuredInstrumentationKey];
  delete global[directInstrumentationKey];
  delete global[defaultOTelInstrumentationKey];
}

function defaultOTelInstrumentation(): OTelInstrumentation {
  if (!global[defaultOTelInstrumentationKey]) {
    global[defaultOTelInstrumentationKey] = new OTelInstrumentation();
  }
  return global[defaultOTelInstrumentationKey];
}

/**
 * Lazily creates the dev instrumentation. Imported dynamically only when a
 * telemetry server is configured so the Direct sink stays out of prod paths.
 */
function directInstrumentation(): Instrumentation | undefined {
  const serverConfigured =
    !!telemetryServerUrl || !!process.env.GENKIT_TELEMETRY_SERVER;
  if (!serverConfigured) return undefined;
  if (!global[directInstrumentationKey]) {
    global[directInstrumentationKey] = new DirectTelemetryInstrumentation();
  }
  return global[directInstrumentationKey];
}

/**
 * Resolves the active instrumentation chain for this call.
 * - base: `[configured]` if `configureInstrumentation` was called, else the
 *   implicit `[OTelInstrumentation]` default.
 * - dev: `DirectTelemetryInstrumentation` is prepended when a telemetry server
 *   is configured, so ids from Direct win and the Dev UI is fed without OTel.
 *
 * @hidden
 */
export function activeInstrumentations(): Instrumentation[] {
  const base: Instrumentation =
    global[configuredInstrumentationKey] ?? defaultOTelInstrumentation();
  const direct = directInstrumentation();
  return direct ? [direct, base] : [base];
}

/** @hidden */
export function isInstrumentedBy(
  ctor: new (...args: any[]) => Instrumentation
): boolean {
  return activeInstrumentations().some((i) => i instanceof ctor);
}

// ---------------------------------------------------------------------------
// Dispatcher.
// ---------------------------------------------------------------------------

/**
 * Runs the provided function in a new span.
 * @deprecated
 * @hidden
 */
export async function runInNewSpan<T>(
  registry: Registry | HasRegistry,
  opts: RunInNewSpanOpts,
  fn: RunInNewSpanFn<T>
): Promise<T>;

/**
 * Runs the provided function in a new span.
 * @hidden
 */
export async function runInNewSpan<T>(
  opts: RunInNewSpanOpts,
  fn: RunInNewSpanFn<T>
): Promise<T>;

/**
 * Runs the provided function in a new span.
 * @hidden
 */
export async function runInNewSpan<T>(
  registryOrOprs: Registry | HasRegistry | RunInNewSpanOpts,
  optsOrFn: RunInNewSpanOpts | RunInNewSpanFn<T>,
  fnMaybe?: RunInNewSpanFn<T>
): Promise<T> {
  let opts: RunInNewSpanOpts;
  let fn: RunInNewSpanFn<T>;
  if (arguments.length === 3) {
    opts = optsOrFn as RunInNewSpanOpts;
    fn = fnMaybe as RunInNewSpanFn<T>;
  } else {
    opts = registryOrOprs as RunInNewSpanOpts;
    fn = optsOrFn as RunInNewSpanFn<T>;
  }
  await ensureBasicTelemetryInstrumentation();

  const parentStep =
    getAsyncContext().getStore<SpanContext>(spanMetadataAlsKey);
  if (!parentStep) opts.metadata.isRoot ||= true;

  // Genkit-semantic setup (backend independent). Providers only encode the
  // metadata/labels we build here.
  opts.metadata.path = buildPath(
    opts.metadata.name,
    parentStep?.metadata?.path || '',
    opts.labels
  );

  const isGenkitSpan = !!opts.labels?.[SPAN_TYPE_ATTR];
  const labels = { ...opts.labels };
  if (isGenkitSpan && parentStep) {
    const parentIsGenkit = !!parentStep.labels?.[SPAN_TYPE_ATTR];
    if (!parentIsGenkit && parentStep.spanId) {
      // Stitch orphaned genkit spans back to their nearest genkit ancestor.
      // Encoded as a label so every provider records it identically.
      labels[ATTR_PREFIX + ':lastKnownParentSpanId'] = parentStep.spanId;
    }
  }

  const spanContext = {
    ...parentStep,
    metadata: opts.metadata,
    labels,
  } as SpanContext;

  const info: InstrumentationSpanInfo = {
    metadata: opts.metadata,
    labels,
    links: opts.links,
  };

  // Innermost callback: every provider span/context is collected by now, so we
  // resolve the composite ids, seed the ALS store for children, run `fn`, and
  // set success/error state before the providers finalize their spans.
  const runBody = async (contexts: GenkitSpanContext[]): Promise<T> => {
    const compositeCtx = makeCompositeContext(contexts);
    if (isGenkitSpan) {
      spanContext.spanId = compositeCtx.spanId || undefined;
    }
    const compositeSpan = makeCompositeSpan(compositeCtx);

    return await getAsyncContext().run(spanMetadataAlsKey, spanContext, () =>
      runWithGenkitState(opts.metadata, spanContext, compositeSpan, fn)
    );
  };

  return await dispatch(activeInstrumentations(), info, runBody);
}

/**
 * Nests each provider's `runInNewSpan` (outermost first), collecting the
 * context each hands to `next`, and invokes `runBody` at the center.
 */
async function dispatch<T>(
  chain: Instrumentation[],
  info: InstrumentationSpanInfo,
  runBody: (contexts: GenkitSpanContext[]) => Promise<T>
): Promise<T> {
  const contexts: GenkitSpanContext[] = [];

  const at =
    (i: number) =>
    async (_span: ApiSpan, ctx: GenkitSpanContext): Promise<T> => {
      contexts[i] = ctx;
      if (i + 1 < chain.length) {
        return chain[i + 1].runInNewSpan(info, at(i + 1));
      }
      return runBody(contexts);
    };

  return chain[0].runInNewSpan(info, at(0));
}

/**
 * Genkit success/error bookkeeping that used to live inline in `runInNewSpan`.
 * Runs around `fn`; providers see the rethrown error and finalize afterwards.
 */
async function runWithGenkitState<T>(
  metadata: SpanMetadata,
  spanContext: SpanContext,
  compositeSpan: ApiSpan,
  fn: RunInNewSpanFn<T>
): Promise<T> {
  try {
    const output = await fn(metadata, compositeSpan);
    if (metadata.state !== 'error') {
      metadata.state = 'success';
    }
    recordPath(metadata, spanContext);
    return output;
  } catch (e) {
    recordPath(metadata, spanContext, e);
    metadata.state = 'error';
    // Mark the first failing span as the source of failure. Prevent parent
    // spans that catch re-thrown exceptions from also claiming to be the
    // source.
    if (typeof e === 'object' && e !== null) {
      if (!(e as any).ignoreFailedSpan) {
        metadata.isFailureSource = true;
      }
      (e as any).ignoreFailedSpan = true;
    }
    throw e;
  }
}

/**
 * The span handed to `fn`. `spanContext()` returns the composite-resolved ids;
 * `fn` callers only ever read those. Attribute writes fan out to providers via
 * the composite context.
 */
function makeCompositeSpan(ctx: GenkitSpanContext): ApiSpan {
  // Non-recording wrapper around the composite ids. Type-honest (real ApiSpan),
  // and does not boot the OTel SDK.
  return trace.wrapSpanContext({
    traceId: ctx.traceId || '0'.repeat(32),
    spanId: ctx.spanId || '0'.repeat(16),
    traceFlags: TraceFlags.SAMPLED,
  });
}

function makeCompositeContext(
  contexts: GenkitSpanContext[]
): GenkitSpanContext {
  // A third-party provider may call next() without a ctx, so guard each access.
  return {
    get traceId() {
      return firstNonEmpty(contexts.map((c) => c?.traceId));
    },
    get spanId() {
      return firstNonEmpty(contexts.map((c) => c?.spanId));
    },
    setMetadata(values: Record<string, unknown>) {
      for (const c of contexts) c?.setMetadata(values);
    },
  };
}

/**
 * Creates a new child span and attaches it to a previously created trace. This
 * is useful, for example, for adding deferred user engagement metadata.
 *
 * @hidden
 */
export async function appendSpan(
  traceId: string,
  parentSpanId: string,
  metadata: SpanMetadata,
  labels?: Record<string, string>
) {
  await ensureBasicTelemetryInstrumentation();

  const tracer = trace.getTracer(TRACER_NAME, TRACER_VERSION);

  const spanContext = trace.setSpanContext(ROOT_CONTEXT, {
    traceId: traceId,
    traceFlags: 1, // sampled
    spanId: parentSpanId,
  });

  // TODO(abrook): add explicit start time to align with parent
  const span = tracer.startSpan(metadata.name, {}, spanContext);
  span.setAttributes(metadataToAttributes(metadata));
  if (labels) {
    span.setAttributes(labels);
  }
  span.end();
}

/**
 * Sets provided attribute value in the current span.
 *
 * @hidden
 */
export function setCustomMetadataAttribute(key: string, value: string) {
  const currentStep = getCurrentSpan();
  if (!currentStep) {
    return;
  }
  if (!currentStep.metadata) {
    currentStep.metadata = {};
  }
  currentStep.metadata[key] = value;
}

/**
 * Sets provided attribute values in the current span.
 *
 * @hidden
 */
export function setCustomMetadataAttributes(values: Record<string, string>) {
  const currentStep = getCurrentSpan();
  if (!currentStep) {
    return;
  }
  if (!currentStep.metadata) {
    currentStep.metadata = {};
  }
  for (const [key, value] of Object.entries(values)) {
    currentStep.metadata[key] = value;
  }
}

/**
 * Converts a fully annotated path to a friendly display version for logs
 *
 * @hidden
 */
export function toDisplayPath(path: string): string {
  const pathPartRegex = /\{([^\,}]+),[^\}]+\}/g;
  return Array.from(path.matchAll(pathPartRegex), (m) => m[1]).join(' > ');
}

function getCurrentSpan(): SpanMetadata {
  const step = getAsyncContext().getStore<SpanContext>(spanMetadataAlsKey);
  if (!step) {
    throw new Error('running outside step context');
  }
  return step.metadata;
}

function buildPath(
  name: string,
  parentPath: string,
  labels?: Record<string, string>
) {
  const stepType =
    labels && labels['genkit:type']
      ? `,t:${labels['genkit:metadata:subtype'] === 'flow' ? 'flow' : labels['genkit:type']}`
      : '';
  return parentPath + `/{${name}${stepType}}`;
}

function recordPath(
  spanMeta: SpanMetadata,
  spanContext: SpanContext,
  err?: any
) {
  const path = spanMeta.path || '';
  const decoratedPath = decoratePathWithSubtype(spanMeta);
  // Only add the path if a child has not already been added. In the event that
  // an error is rethrown, we don't want to add each step in the unwind.
  const paths = Array.from(spanContext?.paths || new Set<PathMetadata>());
  const status = err ? 'failure' : 'success';
  if (!paths.some((p) => p.path.startsWith(path) && p.status === status)) {
    const now = performance.now();
    const start = spanContext?.timestamp || now;
    spanContext?.paths?.add({
      path: decoratedPath,
      error: err?.name,
      latency: now - start,
      status,
    });
  }
  spanMeta.path = decoratedPath;
}

function decoratePathWithSubtype(metadata: SpanMetadata): string {
  if (!metadata.path) {
    return '';
  }

  const pathComponents = metadata.path.split('}/{');

  if (pathComponents.length == 1) {
    return metadata.path;
  }

  const stepSubtype =
    metadata.metadata && metadata.metadata['subtype']
      ? `,s:${metadata.metadata['subtype']}`
      : '';
  const root = `${pathComponents.slice(0, -1).join('}/{')}}/`;
  const decoratedStep = `{${pathComponents.at(-1)?.slice(0, -1)}${stepSubtype}}`;
  return root + decoratedStep;
}

File: go/core/tracing/instrumentation.go
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package tracing

import (
	"context"
	"maps"
	"sync"
	"sync/atomic"
)

// This file holds the pluggable instrumentation abstraction: span creation is
// decoupled from collection/export. [RunInNewSpan] is the entry point and owns
// Genkit semantics (path, isRoot, state/output, debug logging); it dispatches
// to a chain of Instrumentation providers that each only encode the span into
// their backend.

// SpanInfo is the backend-independent view of a span the dispatcher hands to
// each [Instrumentation] provider. The dispatcher owns Genkit semantics; a
// provider only encodes this into its backend, reading it through Labels and
// the accessor methods.
//
// Everything the accessors return is known when the span starts. The run's
// result is not on SpanInfo: a provider reads the output and error from what
// next returns, the only point they are known.
type SpanInfo struct {
	// Labels are the raw TelemetryLabels set directly as span attributes.
	Labels map[string]string

	// metadata is the dispatcher-owned running metadata. In-package providers
	// (OTel, Direct) encode it via its attributes()/startAttributes() methods.
	// It is fully populated (state, output, error) only after next returns.
	metadata *spanMetadata
}

// The accessors tolerate a nil receiver and nil metadata, so a provider (or a
// test) holding a zero-value *SpanInfo does not panic.

// Name is the span name. For a model span it is the fully qualified model name
// (e.g. "googleai/gemini-flash-latest").
func (i *SpanInfo) Name() string {
	if i == nil || i.metadata == nil {
		return ""
	}
	return i.metadata.Name
}

// Type is the Genkit span type ("action", "flowStep", "util", ...).
func (i *SpanInfo) Type() string {
	if i == nil || i.metadata == nil {
		return ""
	}
	return i.metadata.Type
}

// Subtype is the finer categorization ("model", "tool", "flow", ...), or "".
func (i *SpanInfo) Subtype() string {
	if i == nil || i.metadata == nil {
		return ""
	}
	return i.metadata.Subtype
}

// Path is the type-annotated span path, e.g.
// "/{chatFlow,t:flow}/{googleai/gemini-flash-latest,t:action,s:model}".
func (i *SpanInfo) Path() string {
	if i == nil || i.metadata == nil {
		return ""
	}
	return i.metadata.Path
}

// IsRoot reports whether this span is the root of a Genkit trace.
func (i *SpanInfo) IsRoot() bool {
	if i == nil || i.metadata == nil {
		return false
	}
	return i.metadata.IsRoot
}

// Input is the raw input the action was invoked with.
func (i *SpanInfo) Input() any {
	if i == nil || i.metadata == nil {
		return nil
	}
	return i.metadata.Input
}

// Metadata returns a copy of the span's custom metadata (recorded as
// genkit:metadata:<key> attributes), or nil if there is none. Metadata added
// mid-run arrives through [Span.SetMetadata] instead.
func (i *SpanInfo) Metadata() map[string]string {
	if i == nil || i.metadata == nil {
		return nil
	}
	return maps.Clone(i.metadata.Metadata)
}

// Span is the backend-independent handle a provider exposes for the span it
// opened. The dispatcher reads TraceInfo to build the composite ids and fans
// SetMetadata out for mid-run custom metadata.
type Span interface {
	// TraceInfo returns the provider's ids, or empty strings if it tracks none.
	TraceInfo() TraceInfo
	// SetMetadata records mid-run custom metadata on the span.
	SetMetadata(md map[string]string)
}

// NextFunc continues the instrumentation chain. A provider must call it exactly
// once, passing the context children should observe and the Span it opened, and
// return its result.
type NextFunc func(ctx context.Context, span Span) (any, error)

// Instrumentation is a pluggable span-creation provider: middleware over span
// creation. Implementations open a backend span, call next, and finalize from
// info once it returns.
type Instrumentation interface {
	RunInNewSpan(ctx context.Context, info *SpanInfo, next NextFunc) (any, error)
}

// ---------------------------------------------------------------------------
// Provider registry.
// ---------------------------------------------------------------------------

var (
	// instrumentationMu serializes writers of the registry; readers on the
	// span hot path only load activeChain.
	instrumentationMu sync.Mutex
	// configured replaces the implicit OTel default when set via
	// ConfigureInstrumentation.
	configured Instrumentation
	// defaultOTel is the implicit, back-compat default. Removed in the next
	// major to reach the shared "not instrumented by default" goal.
	defaultOTel Instrumentation = &OTelInstrumentation{}
	// direct feeds the Dev UI without OTel; set by EnableDevInstrumentation.
	direct Instrumentation
	// activeChain is the resolved chain every span dispatches through,
	// rebuilt by each registry write.
	activeChain atomic.Pointer[[]Instrumentation]
)

func init() {
	rebuildChainLocked()
}

// ConfigureInstrumentation replaces the implicit default
// ([OTelInstrumentation]) with i. DirectTelemetryInstrumentation is still
// prepended in dev (when a telemetry server is configured), so a typical dev
// chain becomes [Direct, i].
func ConfigureInstrumentation(i Instrumentation) {
	instrumentationMu.Lock()
	defer instrumentationMu.Unlock()
	configured = i
	rebuildChainLocked()
}

// EnableDevInstrumentation installs the Direct provider, feeding the Dev UI's
// telemetry server at url, at the front of the chain. An empty url is ignored;
// a later call updates the destination. In the dev environment genkit.Init
// calls it with GENKIT_TELEMETRY_SERVER, and the reflection server calls it
// with the URL the Genkit CLI supplies.
func EnableDevInstrumentation(url string) {
	if url == "" {
		return
	}
	instrumentationMu.Lock()
	defer instrumentationMu.Unlock()
	direct = newDirectTelemetryInstrumentation(url)
	rebuildChainLocked()
}

// ResetInstrumentation clears configured and dev instrumentation, restoring
// the implicit OTel default. Intended for tests that call
// [ConfigureInstrumentation] and want to undo it (typically via t.Cleanup).
func ResetInstrumentation() {
	instrumentationMu.Lock()
	defer instrumentationMu.Unlock()
	configured = nil
	direct = nil
	rebuildChainLocked()
}

// activeInstrumentations returns the chain for a new span: the Direct provider
// first when a dev telemetry server is configured (so its ids win), then the
// configured provider or the implicit OTel default. Lock-free: this runs for
// every span.
func activeInstrumentations() []Instrumentation {
	return *activeChain.Load()
}

// rebuildChainLocked publishes the chain for the current registry state.
// Caller holds instrumentationMu (or is init).
func rebuildChainLocked() {
	base := configured
	if base == nil {
		base = defaultOTel
	}
	chain := []Instrumentation{base}
	if direct != nil {
		chain = []Instrumentation{direct, base}
	}
	activeChain.Store(&chain)
}

// ---------------------------------------------------------------------------
// Dispatcher.
// ---------------------------------------------------------------------------

// dispatch nests each provider's RunInNewSpan (outermost first), collecting the
// Span each hands to next, and invokes runBody at the center with a composite
// span over them all.
func dispatch(ctx context.Context, chain []Instrumentation, info *SpanInfo, runBody NextFunc) (any, error) {
	spans := make([]Span, len(chain))
	var at func(i int) NextFunc
	at = func(i int) NextFunc {
		return func(ctx context.Context, span Span) (any, error) {
			spans[i] = span
			if i+1 < len(chain) {
				return chain[i+1].RunInNewSpan(ctx, info, at(i+1))
			}
			return runBody(ctx, compositeSpan(spans))
		}
	}
	return chain[0].RunInNewSpan(ctx, info, at(0))
}

// compositeSpan resolves TraceInfo as the first non-empty ids across the chain
// (Direct wins in dev) and fans SetMetadata out to every provider.
type compositeSpan []Span

func (cs compositeSpan) TraceInfo() TraceInfo {
	var out TraceInfo
	// A third-party provider may call next with a nil span, so guard each.
	for _, s := range cs {
		if s == nil {
			continue
		}
		ti := s.TraceInfo()
		if out.TraceID == "" && ti.TraceID != "" {
			out.TraceID = ti.TraceID
		}
		if out.SpanID == "" && ti.SpanID != "" {
			out.SpanID = ti.SpanID
		}
	}
	return out
}

func (cs compositeSpan) SetMetadata(md map[string]string) {
	for _, s := range cs {
		if s != nil {
			s.SetMetadata(md)
		}
	}
}
